docker/indexer/stages/preparation/preparation.go

```diff
@@ -0,0 +1,294 @@
+// Package preparation provides functionality to extract tags, branches and commits from repository configurations.
+package preparation
+
+import (
+	"archive/tar"
+	"context"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/google/osv.dev/docker/indexer/shared"
+
+	log "github.com/golang/glog"
+	pb "github.com/google/osv.dev/docker/indexer/proto"
+)
+
+const workers = 5
+
+// Result is the data structure returned by the stage.
+type Result struct {
+	Name            string
+	BaseCPE         string
+	VersionRE       *regexp.Regexp
+	TagMatching     bool
```

Comment:
When would this ever be not true?

Reply:
Not sure I'm following you on this. Some repos do tag based versioning others based on branch names and then some only use commit hashes.

We could also decide to just hash everything and do a best effort detection of the version as you mentioned in a later comment. Thus eliminating the need for TagMatching, BranchMatching and HashAllCommits.

Reply:
Sorry for being unclear. I think TagMatching would be the most common, and it seems very unlikely this would ever need to be true. Even if tags aren't used, by either using the version regex specified or doing the best effort detection we can skip tags that we can't get extract version information from. 

This can be extended to branch matching too, but I'm unsure we can always just do HashAllCommits because repos can get very big and that would get slow very quickly.